yt/utilities/lib/cyoctree.pyx

```diff
@@ -0,0 +1,590 @@
+#
+# To maintain backwards compatibility we also have the cell structure. This
+# means that each leaf is split into 2**(3*over_refine_factor) cells. When an
+# SPH field is interpolated onto the octree, we interpolate the value are the
+# centre of each cell. The cell locations, and particles they contain, are *NOT*
+# stored in memory instead this is calculated on the fly when a interpolation or
+# cell position request is made.
+# This adds a slight overhead for these functions, but reduces the memory usage
+# by 2**(3*over_refine_factor)
+
+cdef struct Node:
+    double left_edge[3]
+    double right_edge[3]
+
+    np.int64_t start                #  star and end of particles we store
+    np.int64_t end
+
+    np.int64_t parent               # position of parent in Octree.nodes
+    np.int64_t children             # position of 0th child, children are
+                                    # contiguous
+    bool leaf
+    np.int64_t node_id              # not sure if this is even useful
+    np.int64_t leaf_id
+    unsigned char depth
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef class CyOctree:
+    cdef double _left_edge[3]       # boundary conditions for the octree
+    cdef double _right_edge[3]
+    cdef np.int64_t[:] _idx         # ordering of particles used by the tree
+
+    cdef int _n_ref                 # number of particles per leaf
+    cdef np.int64_t _num_particles
+
+    cdef np.int64_t _data_version
+
+    cdef vector[Node] nodes         # This is an STL container to store the octs
+
+    # Cell structure
+    cdef int _max_depth
+    cdef int _over_refine_factor    # this allows the tree to be built with more
+                                    # than 8 cells per leaf
+    cdef int _num_cells             # 2**(3*_over_refine_factor)
+    cdef int _num_cells_per_dim     # 2**(_over_refine_factor)
+
+    # Children structure
+    cdef int _density_factor        # this allows the tree to be built with more
+                                    # than 8 children per node
+    cdef int _num_children          # 2**(3*_density_factor)
+    cdef int _num_children_per_dim  # 2**(_density_factor)
+
+    # this is use for interpolation and is global for the octree smoothing
+    # operations
+    cdef kernel_func kernel
+
+    def __init__(self, double[:, ::1] &input_pos = None, left_edge = None,
+                 right_edge = None, int n_ref=32, int over_refine_factor=1,
+                 int density_factor=1, np.int64_t data_version=0,
+                 int max_depth=20):
+
+        # if this is the case, we are very likely just initialising an instance
+        # and then going to load an existing Octree from memory, so we don't
+        # really need to do anything
+        if input_pos is None:
+            return
+
+        # These don't have setters as these would invalidate the tree which is
+        # a feature we don't have
```

Comment:
but I plan to add, if we have an octree with n_ref = 32 and someone does,
`octree.n_ref = ..`
it would be nice to just either continue refining or delete some octs dependent on whether the new `n_ref` is larger or smaller